### ISSUE TITLE
What direction is Symfony heading in? 

### ISSUE THREAD

**Issue by menzow** [symfony/symfony#19557]
Hello,

After almost 1.5 years of not using Symfony, I have a new project that I would like to use symfony for. But after reading up on the best practices for 3.x, and looking at the directory structure I noticed some differences. 

For embedding assets into your project, it is now advised to use the `/web` directory "because it makes the lives of designers and frontend developers easier" .
Now what I remember is that Bundles should be setup in a way that they can run independently from each other or maybe depend loosely on one another.

Now if I have a bundle that has some custom CSS, that CSS must by put ( according to best practices ) in the Web directory. Making it very difficult to properly deploy that Bundle into multiple projects.

I also noticed a /tests directory in the root of the project. With a default layout of `/tests/AppBundle/Controller/DefaultControllerTest.php`.

Why are tests not in the same directory as the /src/AppBundle/X directory?

Even the templates for the AppBundle are again placed in the /app/Resources/views directory instead of /src/AppBundle/Resources/views.

To me it feels that new developers are giving the instruction to scatter files over the `/app`,`/web`, `/src`, `/tests` directories. Giving a project no clear direction where files are. And projects will not be self contained into one directory.

Why is there no clear single /src/AppBundle/ containing all code required to run the bundle?

What is the idea behind this, and what is the direction syfmony developers see this heading in?

I prefer to have directory structures like this:
```
/src/AppBundle
    /Controller
        ...
    /Entity
        ...
    /Resources
        ...
    ...
```

Or better;
```
/src/AppBundle/
    /User
        /Entity
            UserEntity
        /Resources
            /config
                product.routes.yml
            /public
                /assets
            /views
                user.list.html.twig
                user.edit.html.twig
                user.create.html.twig
        /Controller
            UserViewController.php
            UserListController.php
            UserEditController.php
    /Products
        /Entity
            ProductEntity.php
        /Service
            ProductService.php
        /Controller
            ProductViewController
            ProductListController.php
            ProductEditController.php
        /Resources
            /config
                product.routes.yml
            /views
                product.list.html.twig
                product.edit.html.twig
                product.create.html.twig
```

The last one is purely personal preference, and it doesn't fit anything ever used in Symfony so I understand that something like that would not be introduced. But in the second example, code belonging to anything AppBundle does, is added to the AppBundle directory. Making it easy to re-use in other projects, or remove from the current project entirely.

**Comment by ro0NL** [comment]
Relax! It's all good ;-) However i strongly recommend you to read some [documentation](https://symfony.com/doc/current/index.html).

For assets;
You can still put them in your bundles, its even good practice to keep them together and somehow compiling/making them available in your web directory. For example, look at `bin/console assets:install` or assetic. See also the [frameworkbundle](https://github.com/symfony/symfony/tree/master/src/Symfony/Bundle/FrameworkBundle/Resources/public).

For tests;
Separating source from test files is by convention and basicly makes configuring phpunit easier/more reliable. See also the [frameworkbundle](https://github.com/symfony/symfony/tree/master/src/Symfony/Bundle/FrameworkBundle/Tests), which has still tests in the source directory.

For templates;
Templates in the app directory allows for overriding bundle templates, its optional! Or at least it's used for application specific templates. See also the [frameworkbundle](https://github.com/symfony/symfony/tree/master/src/Symfony/Bundle/FrameworkBundle/Resources/views) and [documentation](https://symfony.com/doc/current/templating/overriding.html).

Long story short; most things can and should live in your bundle structure. For class/file structure you're also free to go, it's just a matter of configuring composer accordingly.

**Comment by ro0NL** [comment]
Your last proposal structure would indeed not work for _symfony_ bundles. You would have a `UserBundle` and `ProductBundle`, living in your app/vendor namespace..

**Comment by menzow** [comment]
@ro0NL Thank you for your response. But I don't think you fully understood my questions.
I did read the documentation, and I know it's possible for templates to be overridden through the /app directory and assets to be placed in the Bundle folder.

My question is about the recommendations given by Symfony. The best practices etc.
If you look [here](http://symfony.com/doc/current/best_practices/web-assets.html) you see that it's best practice to store all your assets in the /web directory.

And [here](http://symfony.com/doc/current/testing.html) they say test should be put into the /tests directory.

Cited from best practices version 2.6

```

When Symfony 2.0 was released, most developers naturally adopted the symfony 1.x way of dividing
applications into logical modules. That's why many Symfony apps use bundles to divide their code into
logical features: UserBundle, ProductBundle, InvoiceBundle, etc.
But a bundle is meant to be something that can be reused as a stand-alone piece of software. If
UserBundle cannot be used "as is" in other Symfony apps, then it shouldn't be its own bundle. Moreover,
if InvoiceBundle depends on ProductBundle, then there's no advantage to having two separate bundles.

```
Now here they say if a Bundle can't be used `as is` than it shouldn't be a bundle.

A bundle needs assets and tests if you want to share it fully. So why is it best practice to put these important parts of code outside of the bundle directory?

**Comment by menzow** [comment]
After reading best practices for 3.1 it's become clear that it's recommended to have only one AppBundle, and that it's not supposed to be shared independently. So you always store the whole project source in git. Making it a lot more logical why you would use /web for your assets and /test for your tests.

**Comment by ro0NL** [comment]
Exactly, the app bundle is not intended to be a reuseable bundle. For reusable bundles you'd still have a separate repository, on what level you're separating concerns is up to you. However it somehow needs to end up in your /web dir to be functional. (compiling, symlinking, or just physical and not being part of bundle).

**Comment by menzow** [comment]
Yes that is correct. I closed the issue, and thanks for your help on clarifying this. :)

**Comment by sstok** [comment]
"If UserBundle cannot be used "as is" in other Symfony apps, then it shouldn't be **its own** bundle." 

Notice the "its own", in the past many developers have created bundles that depended heavily on each other, "InvoiceBundle depends on ProductBundle" but if you only want the InvoiceBundle then you also need the ProductBundle and thus there is coupling between the two.

If the InvoiceBundle can be used alone (without depending on other specific bundles), then it's is own bundle, else InvoiceBundle and ProductBundle should be kept together in say a ShoppingBundle (where they share there code and logic).

If you take this further you can apply the principle of bounded context: Each Bundle has it's own Entities and Models, and no Bundle may depend on the Entities and Models of another bundle. Instead they are connected (rather then glued) together using a bridge that allows both to communicate together without knowing about each other.

For example, I'm building a webhosting management system;

Instead of making the WebhostingBundle depend on the CustomerBundle, the WebhostingBundle provides an `WebhostingAccountOwner` interface that a bridge Bundle can implement to connect a WebhostingAccount to a eg. Customer, Administrator or anything.

You can choose the install WebhostingBundle, and use whatever you like for it's owner.
Yes, this makes Doctrine ORM really hard to use as you no longer can do `$account->getCustomer()->getName()` but in the long run the code is more decoupled and easier to change without breaking the whole system.


The Symfony AppBundle is about keeping all application specific code together, without the intention of ever sharing it with other applications. It's the recommended way for RAD (Rapid Application development) and applications that cannot be split further.

But for big applications that consist of multiple systems (product, invoice, customer, marketing) it's better to keep each system in it's own bundle without depending on the logic of other bundles.
So you can at anytime move them to a smaller application, this approach is known as a Hexagonal Architecture](https://youtu.be/fgQWnglnGeU). 

If you are just getting started with Symfony just use the Best practices, once you gain more experience you can adapt more advanced and decoupled architectures 👍 

NB. When I talk about other bundles this doesn't include the Framework, TwigBundle etc. But Bundles that were specifically made for the application, and are thus not sharable.

**Comment by menzow** [comment]
@sstok thank you for your in depth reply. It goes a little bit out of the scope of what was initially asked, but it might help some people.

I closed this issue because the question I asked has been answered.
